src/aspire/basis/pswf_2d.py

```diff
@@ -1,5 +1,329 @@
+
 from aspire.basis import Basis
+from aspire.basis.pswf_utils import BNMatrix
+from aspire.basis.basis_utils import d_decay_approx_fun, t_radial_part_mat, t_x_mat, t_x_derivative_mat
+from aspire.basis.basis_utils import k_operator, leggauss_0_1
+
+
+logger = logging.getLogger(__name__)
+
+
+class PSWFBasis2D(Basis):
+    """
+    Define a derived class using the Prolate Spheroidal Wave Function (PSWF) basis for mapping 2D images.
+    The numerical evaluation for 2D PSWFs at arbitrary points in the unit disk is based on the direct method
+    described in the papers as below:
+        1) Boris Landa and Yoel Shkolnisky, "Steerable principal components for space-frequency localized images",
+        SIAM J. Imag. Sci. 10, 508-534 (2017).
+        2) Boris Landa and Yoel Shkolnisky, "Approximation scheme for essentially bandlimited and space-concentrated
+        functions on a disk", Appl. Comput. Harmon. Anal. 43, 381-403 (2017).
+        3) Yoel Shkolnisky, "Prolate spheroidal wave functions on a disc-Integration and approximation of
+        two-dimensional bandlimited functions", Appl. Comput. Harmon. Anal. 22, 235-256 (2007).
+    """
+    def __init__(self, size, gamma_truncation=1.0, beta=1.0):
+        """
+        Initial an object for 2D Prolate Spheroidal Wave Function (PSWF) basis expansion using direct method.
+        """
+        self.rcut = size[0] // 2
+        self.gmcut = gamma_truncation
+        self.beta = beta
+        super().__init__(size)
+
+    def _build(self):
+        """
+        Build internal data structures for the direct 2D PSWF method.
+        """
+        logger.info('Expanding 2D images using direct PSWF method.')
+
+        # initial the whole set of PSWF basis functions based on the bandlimit and eps error.
+        self.bandlimit = self.beta * np.pi * self.rcut
+        self.d_vec_all, self.alpha_all, self.lengths = self.init_pswf_func2d(self.bandlimit, eps=np.spacing(1))
+
+        # generate_the 2D grid and corresponding indices inside the disc.
+        self.generate_grid()
+
+        # precompute the basis functions in 2D grids
+        self.precomp()
+
+    def generate_grid(self):
+        """
+        Generate the 2D sampling grid.
+        """
+        if self.nres % 2 == 0:
+            x_1d_grid = range(-self.rcut, self.rcut)
+        else:
+            x_1d_grid = range(-self.rcut, self.rcut + 1)
+        x_2d_grid, y_2d_grid = np.meshgrid(x_1d_grid, x_1d_grid)
+        r_2d_grid = np.sqrt(np.square(x_2d_grid) + np.square(y_2d_grid))
+        points_in_disk = r_2d_grid <= self.rcut
+        x = y_2d_grid[points_in_disk]
+        y = x_2d_grid[points_in_disk]
+        self.r_2d_grid_in_disk = np.sqrt(np.square(x) + np.square(y)) / self.rcut
+        self.theta_2d_grid_in_disk = np.angle(x + 1j * y)
+        self.image_height = len(x_1d_grid)
+        self.points_in_disk = points_in_disk
+        self.points_in_disk_vec = points_in_disk.reshape(self.image_height ** 2)
+
+    def precomp(self):
```

Comment:
It should be a public function return the private data which consistent with all FB methods.

Reply:
I still don't understand. What private data is it returning?

Reply:
My typo. It should be public data. we discuss it during the development FB methods. We want to have the option to redefine the basis such as resolution. I keep it to consistent with FB methods. I can change to private if necessary.

Reply:
Ok I see. So you're saying it can be called if some parameters change to rebuild the basis? I think it's fine to make it private. I don't foresee any users doing this right now.

Reply:
Change it to a private function in the new commit 259994f.